### DISCUSSION TITLE
Auger/Heat Control Question

### DISCUSSION THREAD

**Discussion by nickeeey** [nebhead/PiFire#43] [Q&A] [answered]
Hi, i got a Lousiana Pellet Smoker but the App and control in General is not that good / not reliable and my Fan is spinning the same speed to every set temperature (is that normal?)
Your Project seems to be pretty perfect for me! But there are 2 things i don't get. You got the Fan and Auger to control the temps. And you only control the auger by selecting timings? I thought it is much more complex and i need special timings based on the temp i want to use the grill. So if the grill runs hotter the auger have to put more pellets into the fire or did i miss something in your controls?
You don't got an app (compared to some Blynk / arduino projects). Is the Website working good? But without alerts or am i wrong?

Edit:
I life in Europe so can i change the temp settings from Fahrenheit to degrees?
Or should i need to fork your software?

**Comment by nebhead** [comment, marked as answer]
Hi @nickeeey, 

Welcome to the PiFire discussions.  I'm not really familiar with the internals of the Louisiana Pellet Smoker (or really the app or controls).  However, I can tell you that for Traegers, they don't change the fan speed at all.  They will cycle the fan to increase smoke output, but they don't really use it to control temperature.  

This is the same for PiFire.  The auger is really the main driver for temperature control.  The fan itself we use to cycle, to increase smoke output, but only if desired.  The PID works by running the auger for different amounts of time during a 20 second cycle, which helps to stabilize around a certain set point.  It's by no means perfect, but it works pretty well.  I'm not sure how it would compare to a more modern smoker grill like yours, but for some of us with older Traeger grills, it's beats the old (non-wifi) internal controller that came with it.  

And for the WebUI, it's a matter of personal preference.  I personally like to use the WebUI on both my computer and my phone.  I feel it works well enough for me, and what I need.  However @weberbox has started to create an Android application for PiFire and he has pushed a version of PiFire that supports this in the `development` branch.  I think it looks incredible, and plan to also give it a try.  This might be a little advanced currently, since it requires a little more tinkering with the code.  

BTW - In regards to converting the temperature from F to C, I'm going to work on that in a future update - so it's coming soon hopefully.

**Reply by nickeeey** [reply]
Thanks for your detailed response!
I think these grills works the same way but it doesn't matter, because your control got good results. I just thought it can't be the auger alone based on the p settings. But now I got it. Thanks.
A Celsius option sounds good but maybe I can do it on my own and provide it to you or calculate it for me until you release it.
Thank you. Now I know what I need to know to start the project

**Reply by thomasstrach** [reply]
Hi, I am quite very impressed by this project and imam building it right now for my smoker. 
Is there any update for the switch to Celsius units?

**Reply by nebhead** [reply]
> Hi, I am quite very impressed by this project and imam building it right now for my smoker. Is there any update for the switch to Celsius units?

Thanks Thomas!  My plan is to start working on Celsius in the next few weeks.  I've been delayed by a few other projects, etc. but I'm hoping to tackle this next.  Depending on how difficult it ends up being, it may be sooner or later. :)

**Reply by nebhead** [reply]
@thomasstrach @nickeeey 

FYI, I've started on the changes for supporting Celsius units.  It's pretty close to complete.  My plan is to release the changes to the development branch, hopefully very soon, and let others give it some testing on their systems.   I'll try to drop a pointer to the commit here when it's ready.

**Reply by nebhead** [reply]
@weberbox BTW - While I don't think the addition of Celsius will break the Android Socket IO stuff, it might need an update to support Celsius sometime in the future.  Anyway, I'll let you take a gander once I get the commit posted to the development branch.

**Reply by weberbox** [reply]
> @weberbox BTW - While I don't think the addition of Celsius will break the Android Socket IO stuff, it might need an update to support Celsius sometime in the future.  Anyway, I'll let you take a gander once I get the commit posted to the development branch. 

Ok sounds good. I have already committed some changes to the Android dev branch to add Celsius support. Once you commit the changes I will take a look and see what needs to be changed for Socket IO and/or the the Android app.

**Reply by nebhead** [reply]
Ok, here it is!  I'm sure there are a few bugs in here, but from my initial testing it looks OK.  Proof will be in the testing.  

https://github.com/nebhead/PiFire/commit/32eacc22d23e6f83b23cffca6a24d7a0a6904426

Now it's hopefully just a little more friendly for international users.  (I really wish the US would just switch to Metric & Celsius - life would be just that much easier)

**Reply by weberbox** [reply]
Ok I took a look and this does break the Android app if PiFire is set to use Celsius. The Android app is currently written to use integers for the temps and C is being provided as a float. I will need take a look and see how I can adapt the code to make it work.

**Reply by nebhead** [reply]
Oh, yes, sorry about that.  I just felt that Celsius needed that precision due to the fact that the scale is at least half of what the F scale is.  Thus I made the conscious decision to add the decimal.  I'm open to feedback though.  We can always remove if needed.

**Reply by weberbox** [reply]
> Oh, yes, sorry about that.  I just felt that Celsius needed that precision due to the fact that the scale is at least half of what the F scale is.  Thus I made the conscious decision to add the decimal.  I'm open to feedback though.  We can always remove if needed.  

I think it will be fine. I made some changes last night and testing is going OK so far. I will push the changes to my dev branch once ready. I will also make a pull request to your dev branch to add Socket IO support for changing the units. 

The pull request will also include backup and restore functionality for Socket IO and the Android app as I have already pushed those to my fork of the dev branch.

**Reply by weberbox** [reply]
@nebhead Ok I was about to say it is ready but then I realized that the settings like splus min/max and safety temps are also showing a float value when switching to C. The PiFire settings page does not allow a decimal to be entered but when the switch happens it just converts the F values to C which end up being a float stored in the json. This poses a potential problem with the Android client but I can probably make it work if need be.  

I wanted to ask if you intend on supporting float values in settings or if it would be better to convert/round the temps to int when switching to C since settings does not currently allow floats. Currently it seems the set points are ok as they are still int values and not shown as a float anywhere. 

Edit:
Can you also let me know if your history chart works when switching to C from F? If I do not restart app.py then history gets a 500 error. 

I think the below is supposed to be doing the round but it is still outputting XX.X which is expected with round(x, 1). Was the intent to be a XX.X or just XX? 

```
def convert_temp(units, temp):
	if units == 'F':
		temp_out = int(temp * (9/5) + 32) # Celsius to Fahrenheit
	else:
		temp_out = round((temp - 32) * (5/9), 1) # Fahrenheit to Celcius
	return(temp_out)
```

**Reply by nebhead** [reply]
> I wanted to ask if you intend on supporting float values in settings or if it would be better to convert/round the temps to int when switching to C since settings does not currently allow floats. Currently it seems the set points are ok as they are still int values and not shown as a float anywhere.

Oh, I think we can probably just store those Celsius values in the JSON as int values (without decimals).  That would be much simpler and would make life easier for everyone.  No need to have that kind of fidelity for those settings. :) 
 
> Edit: Can you also let me know if your history chart works when switching to C from F? If I do not restart app.py then history gets a 500 error.

Oh interesting.  It seems to work OK for me, as when you are switching from C to F, it should clear history.  But I'll do a little more testing on that.  

> I think the below is supposed to be doing the round but it is still outputting XX.X which is expected with round(x, 1). Was the intent to be a XX.X or just XX?
> 
> ```
> def convert_temp(units, temp):
> 	if units == 'F':
> 		temp_out = int(temp * (9/5) + 32) # Celsius to Fahrenheit
> 	else:
> 		temp_out = round((temp - 32) * (5/9), 1) # Fahrenheit to Celcius
> 	return(temp_out)
> ```

Right - The intent was to round to the first decimal place.  But in this case, since this function is only used to convert settings in the settings file, we can change that to an int instead of a round.  This will solve the above problem of C settings being stored in the JSON as floats.

**Reply by weberbox** [reply]
> > I wanted to ask if you intend on supporting float values in settings or if it would be better to convert/round the temps to int when switching to C since settings does not currently allow floats. Currently it seems the set points are ok as they are still int values and not shown as a float anywhere.
> 
> Oh, I think we can probably just store those Celsius values in the JSON as int values (without decimals). That would be much simpler and would make life easier for everyone. No need to have that kind of fidelity for those settings. :)
> 
> > Edit: Can you also let me know if your history chart works when switching to C from F? If I do not restart app.py then history gets a 500 error.
> 
> Oh interesting. It seems to work OK for me, as when you are switching from C to F, it should clear history. But I'll do a little more testing on that.
> 
> > I think the below is supposed to be doing the round but it is still outputting XX.X which is expected with round(x, 1). Was the intent to be a XX.X or just XX?
> > ```
> > def convert_temp(units, temp):
> > 	if units == 'F':
> > 		temp_out = int(temp * (9/5) + 32) # Celsius to Fahrenheit
> > 	else:
> > 		temp_out = round((temp - 32) * (5/9), 1) # Fahrenheit to Celcius
> > 	return(temp_out)
> > ```
> 
> Right - The intent was to round to the first decimal place. But in this case, since this function is only used to convert settings in the settings file, we can change that to an int instead of a round. This will solve the above problem of C settings being stored in the JSON as floats.

OK scratch the 500 error for now. I am not sure how I got it to do that as it now seems to be working when switching back and forth. I will watch out for it and let you know if it happens again. It appears for some reason when it was in C it still tried to use the int part of the function from F. Again not sure how I got it to do that as I cannot reproduce now. 

```
[2021-12-08 19:45:47,865] ERROR in app: Exception on /history [GET]
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/flask/app.py", line 2073, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.7/dist-packages/flask/app.py", line 1518, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.7/dist-packages/flask/app.py", line 1516, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.7/dist-packages/flask/app.py", line 1502, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/usr/local/bin/pifire/app.py", line 332, in historypage
    data_blob = prepare_data(num_items, True, settings['history_page']['datapoints'])
  File "/usr/local/bin/pifire/app.py", line 1356, in prepare_data
    data_blob['grill_temp_list'].append(int(data_list[index][1]))
ValueError: invalid literal for int() with base 10: '14.2'
```

**Reply by nebhead** [reply]
Yep - I briefly just reproduced it once, but now I cannot get it to happen again.  :( But anyway, it does seem to be an issue with using int.   I'll also keep an eye out.  Hopefully users won't be switching between C and F frequently. :)

**Reply by thomasstrach** [reply]
just want to let you know that Celsius is working amazing!
Now I have recalculate the American receipt values to Celcsius ;-) 
Everything with the temperature is really working fine, great work also the possibility to switch in the settings.

**Reply by nebhead** [reply]
Glad to hear!  Yeah, it's the recipes that are challenging. :)

**Comment by thomasstrach** [comment]
Hi James,

thanks a lot for your efforts in the Celsius topic here.

 

I have a bit strange behavior in my project:

Everything is running fine based on your great description, I am really surprised about the smooth installation.

There is only one point a bit irregular.

During installation I have selected the default / no hopper sensor selection.

Despite this selection I am getting some %numbers in the display about hopper level.

When I re-configure and choose the Tof sensor it always shows me 100%, even without any sensor.

When I switch back to default / no sensor the value is still alternating with and without the sensor.

Have you seen any behaviour like this?

 

Von: James Weber ***@***.***> 
Gesendet: Mittwoch, 8. Dezember 2021 10:15
An: nebhead/PiFire ***@***.***>
Cc: thomasstrach ***@***.***>; Mention ***@***.***>
Betreff: Re: [nebhead/PiFire] Auger/Heat Control Question (Discussion #43)

 

Ok I took a look and this does break the Android app if PiFire is set to use Celsius. The Android app is currently written to use integers for the temps and C is being provided as a float. I will need take a look and see how I can adapt the code to make it work.

**Reply by nebhead** [reply]
Oh yes, that's due to the prototype module providing random numbers for testing.   If you want, you can manually change this in the `distance_prototype.py` file:

From:
```python
	def GetLevel(self):
		return random.randint(10, 100)
```
To:
```python
	def GetLevel(self):
		return(100)
```

**Reply by weberbox** [reply]
Ya this is my fault. I set that to return a random int for testing when using prototype. It slipped my mind that this is used when someone does not have a distance sensor. When I make the edits to the Socket IO I can add a boolean to check if grillplat is also prototype so it will return random or 100.

**Reply by nickeeey** [reply]
nice to see some updates.
@weberbox are you from Europe? I didnt start the project and need the pcb so maybe we can order together

**Reply by weberbox** [reply]
> nice to see some updates.
> @weberbox are you from Europe? I didnt start the project and need the pcb so maybe we can order together

No I am in the US. I am not sure pricing for Europe but it was only like $15 for 5 pcb's including shipping to the US from JLPCB. Couldn't believe how cheap it is to get custom pcb's now. 

I can't wait to see some of these new designs though. I am probably going to go a little bit a different route as I want the power supply to be part of the board. I have a working setup now so not sure when I will get around to designing a new pcb but I will share it when I do in case anyone wants to use it.

**Reply by nickeeey** [reply]
> > nice to see some updates.
> > @weberbox are you from Europe? I didnt start the project and need the pcb so maybe we can order together
> 
> No I am in the US. I am not sure pricing for Europe but it was only like $15 for 5 pcb's including shipping to the US from JLPCB. Couldn't believe how cheap it is to get custom pcb's now.
> 
> I can't wait to see some of these new designs though. I am probably going to go a little bit a different route as I want the power supply to be part of the board. I have a working setup now so not sure when I will get around to designing a new pcb but I will share it when I do in case anyone wants to use it.

okay.
I dont know why but the lcsc support wont work (or they are out of stock caused by the shortage). And other shops wanted around 40usd+ for 5 Boards but some parts were not automatically recognized and i am not advanced enough to pick the right one
![Uploading Bild_2021-12-08_213845.png…]()

**Reply by weberbox** [reply]
> > > nice to see some updates.
> > > @weberbox are you from Europe? I didnt start the project and need the pcb so maybe we can order together
> > 
> > No I am in the US. I am not sure pricing for Europe but it was only like $15 for 5 pcb's including shipping to the US from JLPCB. Couldn't believe how cheap it is to get custom pcb's now.
> > 
> > I can't wait to see some of these new designs though. I am probably going to go a little bit a different route as I want the power supply to be part of the board. I have a working setup now so not sure when I will get around to designing a new pcb but I will share it when I do in case anyone wants to use it.
> 
> okay.
> I dont know why but the lcsc support wont work (or they are out of stock caused by the shortage). And other shops wanted around 40usd+ for 5 Boards but some parts were not automatically recognized and i am not advanced enough to pick the right one
> ![Uploading Bild_2021-12-08_213845.png…]()
> 

I only order pcb's the other components I get from mouser.com. They are worldwide so might be worth a look. 

Usually you can search the pn# to get a cross reference on Mouser. If it is not found it is just a bit of research to find an alternative.  

I can try to go back and find my order so you can see some part numbers

**Reply by nickeeey** [reply]
okay, now i get it. You order a blank pcb and solder by yourself. Yeah, thats also an option. I mean there are ton of sites with these parts. That should work, but if you can find your order then you can upload it. Thanks

**Reply by nickeeey** [reply]
[bom.xls](https://github.com/nebhead/PiFire/files/7679011/bom.xls)
i tried it with jlc pcb here is a bom with their internal ids. Maybe you can add it to the guide for some people. Its 40USD for 5 Boards

**Reply by nebhead** [reply]
> I can't wait to see some of these new designs though. I am probably going to go a little bit a different route as I want the power supply to be part of the board. I have a working setup now so not sure when I will get around to designing a new pcb but I will share it when I do in case anyone wants to use it.  

Yes!  I really wish EasyEDA had some collaboration options because I think it would be fun to actively collaborate on a design.  Anyway, if you do take a crack at it with an integrated power supply, I will happily test the design for you. :)

**Reply by weberbox** [reply]
> okay, now i get it. You order a blank pcb and solder by yourself. Yeah, thats also an option. I mean there are ton of sites with these parts. That should work, but if you can find your order then you can upload it. Thanks

Here is the order. You do not need line # 5 as line # 4 will be good enough. This also includes the molex connectors for the 120v side. Line # 13 is also optional. I did not end up using it. 

![Screenshot from 2021-12-08 17-15-50](https://user-images.githubusercontent.com/313328/145317271-dfb9a891-1883-4b5b-94d3-25ed8c8fc2e5.png)
![Screenshot from 2021-12-08 17-16-09](https://user-images.githubusercontent.com/313328/145317278-26e13252-2225-4549-bf3e-93a3e84f520f.png)
![Screenshot from 2021-12-08 17-16-28](https://user-images.githubusercontent.com/313328/145317287-f641ca3f-ae8a-48bb-92fc-cdf89d20c9e1.png)

**Reply by nebhead** [reply]
BTW - Someone else just asked how to order a PCB, so I answered in that thread with more specific instructions (and pictures) on how to order.  Link here: [https://github.com/nebhead/PiFire/discussions/58#discussioncomment-1776147](https://github.com/nebhead/PiFire/discussions/58#discussioncomment-1776147)

For Reference: 
* @nebhead Original PCB Design - [https://easyeda.com/nebhead77/pifire-v2](https://easyeda.com/nebhead77/pifire-v2)
* @weberbox Updated PCB Design - [https://easyeda.com/zipster85/pifire-controller-hat](https://easyeda.com/zipster85/pifire-controller-hat)
* @weberbox Button PCB Design (highly recommended) - [https://easyeda.com/zipster85/pifire-controller-hat](https://easyeda.com/zipster85/pifire-controller-hat)
